sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala

```diff
@@ -106,39 +165,66 @@ class BulkWriter(container: CosmosAsyncContainer,
         throw new RuntimeException(s"${writeConfig.itemWriteStrategy} not supported")
     }
 
-    activeTasks.incrementAndGet()
+    semaphore.acquire()
```

Comment:
Wouldn't the retry attempt to acquire again and we are not releasing?

Reply:
1. operation comes, acquires semaphore
2. fails with a retriable status
3. does not release semaphore (we only do it if it `!isGettingRetried`)
4. `scheduleWriteInternal` is called again, acquires semaphore again

Reply:
Maybe the acquire needs to be moved to `scheduleWrite`?

Reply:
very good point Matias. No retry will not re-acquire a semaphore token.

The code is written such a way that the semaphore is released only when a retry doesn't happen (either on success or a final failure). Retry will re-use the same acquired semaphore token.

Reply:
you are right. I meant to move it there. somehow I missed that.

Reply:
thanks. done.